# New Year formula calculator

Calculates a formula of the form (10 * 9) + 8 - 7 + 6 + 5 + 4 - 3 / 2 - 1 using the numbers 10 through 1 and any number of brackets

ibase calculates an integer on base n. We use this in base 4 to generate all of the combinations of operators (+, -, *, /)

In [44]:
import math

def ibase(n, radix=2, maxlen=None):
    r = []
    while n:
        n,p = divmod(n, radix)
        r.append('%d' % p)
        if maxlen and len(r) > maxlen:
            break
    r.reverse()
    return ''.join(r)

def leftpad(s, l):
    while len(s)<l:
        s = "0" + s
    return s

def padbase(n, b, l):
    return leftpad(ibase(n, b), l)


itertools.permutations generates all of the permutations of calculation orders. For a problem with 9 operators there are 9! calculation orders, though many will be redundant

In [45]:
import itertools
p = list(itertools.permutations(range(9)))

print (p[333])
print (math.factorial(9))

(0, 1, 2, 5, 7, 8, 4, 6, 3)
362880


For a given set of operators and calculation order this function derives the answer

In [46]:
def eval_year(opstr, perm, numOperations, inputList):
    
    w = inputList
    div0 = False
    
    for r in range(numOperations):
        for p in perm:
            if p == r:
                
                pos = perm.index(p)
                
                if opstr[pos]=="0":
                    w = w[0:pos] + [ w[pos] + w[pos+1] ] + w[pos+2:]
                elif opstr[pos]=="1":
                    w = w[0:pos] + [ w[pos] - w[pos+1] ] + w[pos+2:]
                elif opstr[pos]=="2":
                    w = w[0:pos] + [ w[pos] * w[pos+1] ] + w[pos+2:]
                elif opstr[pos]=="3":
                    if w[pos+1]==0:
                        w = w[0:pos] + [ w[pos] * w[pos+1] ] + w[pos+2:]
                        div0 = True
                    else:
                        w = w[0:pos] + [ w[pos] / w[pos+1] ] + w[pos+2:]
                elif opstr[pos]=="4":
                    w = w[0:pos] + [ w[pos] ** w[pos+1] ] + w[pos+2:]
                    
                opstr = opstr[0:pos] + opstr[pos+1:]
                perm.remove(p)
                
    if div0 == True:
        ret = 0
    else:
        ret = w[0]
    
    return ret


This function presents the output in a slightly more readable form, inserting brackets to show calculation order. It adds a pair of brackets for every operator, even when not required

In [48]:
def format_output(res, opstr, perm, i, numOperations):
    s = ""
    
    for p in range(numOperations):
        
        pos = perm.index(p)
        
        if opstr[pos]=="0":
            x = "+"
        elif opstr[pos]=="1":
            x = "-"
        elif opstr[pos]=="2":
            x = "*"
        elif opstr[pos]=="3":
            x = "/"
        else:
            x = "**"
        
        i = i[0:pos] + [ "(" + str(i[pos]) + " " + x + " " + str(i[pos+1]) + ") " ] + i[pos+2:]
        opstr = opstr[0:pos] + opstr[pos+1:]
        
        perm.remove(p)
        
        
    return i[0] + " = " + str(res) 
        

Main part of code, where:
- i is the list of numbers in the formula
- numOperators is the number of operators up to four
- numOperations is by definition one less than the number of numbers
- permlist is a list of all of the possible calculation orders (numOperations!)

The total number of possible formulas is therefore numOperations! * numOperators ** numOperations
e.g. 9! * 4^9 = 362880 * 262144 = 95126814720

If any of these evaluate to the target answer we format and print it

In [49]:
import itertools

i = [10,9,8,7,6,5,4,3,2,1]
numOperators = 4
numOperations = len(i)-1
numPermutations = math.factorial(numOperations)

tgt = 2021
res = 0
cnt = 0

permlist = list(itertools.permutations(range(numOperations)))
#permlist = [(0,1,2,3,4,5,6,7,8)]
#generate operators (0,262144)
for operators in range(0, numOperators ** numOperations ):
    
    opstr = padbase (operators, numOperators, numOperations)
    #print (opstr)
    
    #generate calculation orders
    for perm in permlist:
        
        res = eval_year(opstr, list(perm), numOperations, i)
        
        if round(res,4) == tgt:
            #print ("found", res)
            #print (opstr)
            #print (perm)
            print (format_output(res, opstr, list(perm), i, numOperations))
            
    cnt +=1

    if cnt % 1000 == 0:
        print (cnt)
    



1000
2000
3000
4000
5000
6000
7000
8000
9000
10000
11000
12000
13000
14000
15000
16000
17000
18000
19000
20000
21000
22000
23000
24000
25000
26000
27000
28000
29000
30000
31000
32000
33000
34000
35000
36000
37000
38000
39000
40000
41000
42000
43000
44000
45000
46000
47000
48000
49000
50000
51000
52000
53000
54000
55000
56000
57000
58000
59000
60000
61000
62000
63000
64000
65000
66000
67000
68000
69000
70000
71000
72000
73000
74000
75000
76000
77000
78000
79000
80000
81000
82000
83000
84000
85000
86000
87000
88000
89000
90000
91000
92000
93000
94000
95000
96000
97000
98000
99000
100000
101000
102000
103000
104000
105000
106000
107000
108000
(((((((((10 - 9)  * 8)  * 7)  * 6)  + 5)  - 4)  * 3)  * 2)  - 1)  = 2021
109000
110000
111000
112000
113000
114000
115000
116000
117000
118000
119000
120000
121000
122000
123000
124000
125000
126000
127000
128000
129000
130000
131000
132000
133000
134000
135000
136000
137000
138000
139000
140000
141000
142000
143000
144000
145000
146000
147000
148000

In [176]:
eval_year("000020",[0,1,5,2,4,3],numOperations,[9, 8, 7, 6, 5, 4, 3])

000020
[0, 1, 5, 2, 4, 3]
[9, 8, 7, 6, 5, 4, 3]
00020
[1, 5, 2, 4, 3]
[17, 7, 6, 5, 4, 3]
0020
[5, 2, 4, 3]
[24, 6, 5, 4, 3]
020
[5, 4, 3]
[24, 11, 4, 3]
02
[5, 4]
[24, 11, 7]
0
[5]
[24, 77]


101

In [ ]:
w = [10,9,8,7,6,5,4,3,2,1]

n=3

print (w[0:n] +  [w[n]*w[n+1]] + w[n+2:])

In [165]:
permlist = list(itertools.permutations(range(numOperations)))
permlist[0]

(0, 1, 2, 3, 4, 5)